[WPE][GTK] Fix several warnings when building for ARMv7 (32-bits)

clopez opened the pull request:
#### 8e78751b47200105b5533f13fdf31339351c1b21
<pre>
[WPE][GTK] Fix several warnings when building for ARMv7 (32-bits)
<a href="https://bugs.webkit.org/show_bug.cgi?id=247873">https://bugs.webkit.org/show_bug.cgi?id=247873</a>

Reviewed by Michael Catanzaro.

This fixes several warnings that appear when building for ARMv7 (32-bits):
 - deprecated-enum-enum-conversion
 - unused-variable
 - unused-result
 - overflow
 - sign-compare

It also disables two warnings for old versions of GCC
(old versions give false positives).

* PerformanceTests/MallocBench/MallocBench/stress_aligned.cpp:
(benchmark_stress_aligned):
* Source/JavaScriptCore/assembler/ARMv7Assembler.h:
(JSC::ARMv7Assembler::revertJumpTo_movT3movtcmpT2):
(JSC::ARMv7Assembler::ARMInstructionFormatter::twoWordOp12Reg4FourFours):
(JSC::ARMv7Assembler::ARMInstructionFormatter::twoWordOp12Reg4Reg4Imm12):
(JSC::ARMv7Assembler::ARMInstructionFormatter::twoWordOp12Reg4Reg4Reg4Imm8):
(JSC::ARMv7Assembler::ARMInstructionFormatter::twoWordOp12Reg40Imm3Reg4Imm20Imm5):
* Source/WebCore/accessibility/atspi/AccessibilityObjectTextAtspi.cpp:
(WebCore::AccessibilityObjectAtspi::characterIndex const):
(WebCore::AccessibilityObjectAtspi::selectionChanged):
* Source/WebCore/platform/graphics/filters/software/FECompositeSoftwareApplier.cpp:
(WebCore::FECompositeSoftwareApplier::applyArithmetic const):
* Source/WebCore/platform/graphics/gbm/DMABufReleaseFlag.h:
(WebCore::DMABufReleaseFlag::release):
* Source/cmake/WebKitCompilerFlags.cmake:

Canonical link: <a href="https://commits.webkit.org/256792@main">https://commits.webkit.org/256792@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2f3083d6acb8de2d1d3ef02e2ae56d44f4874aa4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/96837 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/6104 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/29934 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/106363 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/166644 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/100815 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/6319 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/34832 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/89219 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/103063 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/102510 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/4732 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/83450 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/31694 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/86568 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/88439 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/74645 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/87783 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/137 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/19922 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/83221 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/125 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/21347 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/28389 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4710 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/4918 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/43867 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/85909 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/1353 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/40641 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19368 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on PerformanceTests/MallocBench/MallocBench/stress_aligned.cpp:
```diff
@@ -148,7 +148,7 @@ void benchmark_stress_aligned(CommandLine&)
     srandom(1); // For consistency between runs.
 
-    size_t limit = 0x00001ffffffffffful;
+    unsigned long long limit = 0x00001ffffffffffful;
```
I see that size_t was incorrect here, but how about:

```suggestion
    uint64_t limit = 0x00001ffffffffffful;
```

clopez replied:
Look nice, I will change it

mcatanzaro commented on Source/WebCore/platform/graphics/gbm/DMABufReleaseFlag.h:
```diff
@@ -72,7 +72,9 @@ struct DMABufReleaseFlag {
         uint64_t value { 1 };
+IGNORE_WARNINGS_BEGIN("unused-result")
         write(fd.value(), &value, sizeof(uint64_t));
+IGNORE_WARNINGS_END
```
I would check with one of the graphics developers before assuming that ignoring errors is desired here. Same for the released() function above, actually.

Normally it is better to check for an error as usual and WTFLogAlways if the write failed; ignoring errors is almost always a bug. Notably, neither the read() nor the write() is checking for EINTR, which is _usually_ a bug. My usual strategy is (a) handle EINTR first (with a loop), then (b) for any other error, print a warning with WTFLogAlways() to give anyone debugging problems a chance. If different behavior is desired here, there should probably be a comment to explain why.

clopez replied:
ACK. I will print a warning to stdout if the write fails.

mcatanzaro commented on Source/cmake/WebKitCompilerFlags.cmake:
```diff
@@ -128,13 +128,17 @@ if (COMPILER_IS_GCC_OR_CLANG)
+    # Warnings to be enabled only on 64-bits
+    if (CMAKE_SIZEOF_VOID_P EQUAL 8)
+        WEBKIT_PREPEND_GLOBAL_COMPILER_FLAGS(-Wcast-align)
+    endif ()
```
I'm not sure if this warning is really useful at all if we're going to disable it for 32-bit systems. x86 does not care about unaligned accesses, so this warning is really there for ARM, right? Maybe just remove it? The list of extra warnings that we want above and beyond those provided by -Wall should ideally be kept short, after all.

clopez replied:
Good point.
I disabled it because warnings are now fatal by default and there are a few warnings of this which I have no much idea how to fix. But I see now how disabling this warning is not the best idea. 

I guess it will be a better idea to make warnings non-fatal by default for 32-bits maybe.

mcatanzaro commented on Source/JavaScriptCore/assembler/ARMv7Assembler.h:
```diff
@@ -1372,7 +1372,7 @@ class ARMv7Assembler {
             twoWordOp5i6Imm4Reg4EncodedImmFirst(OP_MOVT, hi16),
             twoWordOp5i6Imm4Reg4EncodedImmSecond(right, hi16),
-            static_cast<uint16_t>(OP_CMP_reg_T2 | left)
+            static_cast<uint16_t>(static_cast<uint16_t>(OP_CMP_reg_T2) | static_cast<uint16_t>(left))
```
I missed this earlier, but you can remove the outer cast now:

```suggestion
            static_cast<uint16_t>(OP_CMP_reg_T2) | static_cast<uint16_t>(left)
```

clopez replied:
I tried, but then GCC (12.2.0) warns about a narrowing conversion:

```
Source/JavaScriptCore/assembler/ARMv7Assembler.h:1375:50: warning: narrowing conversion of ‘(int)(((short unsigned int)((int)((uint16_t)JSC::ARMv7Assembler::OP_CMP_reg_T2))) | ((short unsigned int)((int)((uint16_t)left))))’ from ‘int’ to ‘uint16_t’ {aka ‘short unsigned int’} [-Wnarrowing]
 1375 |             static_cast<uint16_t>(OP_CMP_reg_T2) | static_cast<uint16_t>(left)
      |             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~

```

mcatanzaro commented on Source/WebCore/platform/graphics/gbm/DMABufReleaseFlag.h:
```diff
@@ -72,7 +72,8 @@ struct DMABufReleaseFlag {
         uint64_t value { 1 };
-        write(fd.value(), &value, sizeof(uint64_t));
+        if (write(fd.value(), &value, sizeof(value)) != sizeof(value))
+            WTFLogAlways("Error writing to the eventfd at DMABufReleaseFlag");
```
You can `#include <wtf/SafeStrerror.h>`, then:

```suggestion
            WTFLogAlways("Error writing to the eventfd at DMABufReleaseFlag: %s", safeStrError(errno));
```

clopez replied:
Nice stuff! will include it :)

clopez committed:
[WPE][GTK] Fix several warnings when building for ARMv7 (32-bits)
https://bugs.webkit.org/show_bug.cgi?id=247873

Reviewed by Michael Catanzaro.

This fixes several warnings that appear when building for ARMv7 (32-bits):
 - deprecated-enum-enum-conversion
 - unused-variable
 - unused-result
 - overflow
 - sign-compare

It also disables two warnings for old versions of GCC
(old versions give false positives).

* PerformanceTests/MallocBench/MallocBench/stress_aligned.cpp:
(benchmark_stress_aligned):
* Source/JavaScriptCore/assembler/ARMv7Assembler.h:
(JSC::ARMv7Assembler::revertJumpTo_movT3movtcmpT2):
(JSC::ARMv7Assembler::ARMInstructionFormatter::twoWordOp12Reg4FourFours):
(JSC::ARMv7Assembler::ARMInstructionFormatter::twoWordOp12Reg4Reg4Imm12):
(JSC::ARMv7Assembler::ARMInstructionFormatter::twoWordOp12Reg4Reg4Reg4Imm8):
(JSC::ARMv7Assembler::ARMInstructionFormatter::twoWordOp12Reg40Imm3Reg4Imm20Imm5):
* Source/WebCore/accessibility/atspi/AccessibilityObjectTextAtspi.cpp:
(WebCore::AccessibilityObjectAtspi::characterIndex const):
(WebCore::AccessibilityObjectAtspi::selectionChanged):
* Source/WebCore/platform/graphics/filters/software/FECompositeSoftwareApplier.cpp:
(WebCore::FECompositeSoftwareApplier::applyArithmetic const):
* Source/WebCore/platform/graphics/gbm/DMABufReleaseFlag.h:
(WebCore::DMABufReleaseFlag::release):
* Source/cmake/WebKitCompilerFlags.cmake:

Canonical link: https://commits.webkit.org/256792@main
